src/Servers/Kestrel/Core/src/Internal/Http3/Http3FrameWriter.cs

```diff
@@ -0,0 +1,329 @@
+        }
+
+        public ValueTask<FlushResult> FlushAsync(IHttpOutputAborter outputAborter, CancellationToken cancellationToken)
+        {
+            lock (_writeLock)
+            {
+                if (_completed)
+                {
+                    return default;
+                }
+
+                var bytesWritten = _unflushedBytes;
+                _unflushedBytes = 0;
+
+                return _flusher.FlushAsync(_minResponseDataRate, bytesWritten, outputAborter, cancellationToken);
+            }
+        }
+
+        internal void WriteResponseHeaders(int streamId, int statusCode, IHeaderDictionary headers)
+        {
+            lock (_writeLock)
+            {
+                if (_completed)
+                {
+                    return;
+                }
+
+                try
+                {
+                    _outgoingFrame.PrepareHeaders();
+                    var buffer = _headerEncodingBuffer.AsSpan();
+                    var done = _qpackEncoder.BeginEncode(statusCode, EnumerateHeaders(headers), buffer, out var payloadLength);
+                    FinishWritingHeaders(streamId, payloadLength, done);
+                }
+                catch (QPackEncodingException hex)
+                {
+                    _http3Stream.Abort(new ConnectionAbortedException(hex.Message, hex));
+                    throw new InvalidOperationException(hex.Message, hex); // Report the error to the user if this was the first write.
+                }
+            }
+        }
+
+        private void FinishWritingHeaders(int streamId, int payloadLength, bool done)
+        {
+            var buffer = _headerEncodingBuffer.AsSpan();
+            _outgoingFrame.Length = payloadLength;
+
+            WriteHeaderUnsynchronized();
+            _outputWriter.Write(buffer.Slice(0, payloadLength));
+
+            while (!done)
+            {
+                done = _qpackEncoder.Encode(buffer, out payloadLength);
+                _outgoingFrame.Length = payloadLength;
+
+                WriteHeaderUnsynchronized();
+                _outputWriter.Write(buffer.Slice(0, payloadLength));
+            }
+        }
+
+        public void Complete()
+        {
+            lock (_writeLock)
+            {
+                if (_completed)
+                {
+                    return;
+                }
+
+                _completed = true;
+                _outputWriter.Complete();
+            }
+        }
+
+        public void Abort(ConnectionAbortedException error)
+        {
+            lock (_writeLock)
+            {
+                if (_aborted)
+                {
+                    return;
+                }
+
+                _aborted = true;
+                _connectionContext.Abort(error);
+
+                Complete();
+            }
+        }
+
+        private static IEnumerable<KeyValuePair<string, string>> EnumerateHeaders(IHeaderDictionary headers)
```

Comment:
Perf 😢

Reply:
What specifically are you referring to? The yield return?